lib/Optimizer/CodeGen/LowerToQIR.cpp

```diff
@@ -549,75 +542,138 @@ class OneTargetOneParamRewrite : public ConvertOpToLLVMPattern<OP> {
+      FlatSymbolRefAttr packingSymbolRef =
           cudaq::opt::factory::createLLVMFunctionSymbol(
-              qirFunctionName, /*return type=*/LLVM::LLVMVoidType::get(context),
-              std::move(tmpArgTypes), parentModule);
+              cudaq::opt::NVQIRPackSingleQubitInArray,
+              /*return type=*/qubitArrayType, {qubitIndexType}, parentModule);
+
+      // Pack the qubit into the array
+      Value result =
+          rewriter
+              .create<LLVM::CallOp>(loc, qubitArrayType, packingSymbolRef,
+                                    adaptor.getControls().front())
+              .getResult();
+      // The array result is what we want for the function args
+      funcArgs.push_back(result);
+    }
 
-      // Add the target op
-      funcArgs.push_back(adaptor.getTargets().front());
+    // If type is a VeqType, then we're good, just forward to the call op
+    if (numControls == 1 && type.isa<quake::VeqType>()) {
+      if (negatedQubitCtrls)
+        return instOp.emitError("unsupported controlled op " + instName +
+                                " with vector of ctrl qubits");
 
-      // Here we already have and Array*, Qubit*
-      rewriter.replaceOpWithNewOp<LLVM::CallOp>(instOp, TypeRange{},
-                                                instSymbolRef, funcArgs);
-      return success();
+      // Add the control array to the args.
+      funcArgs.push_back(adaptor.getControls().front());
     }
 
-    // If the control is a qubit, we need to pack it into an Array*
-    FlatSymbolRefAttr packingSymbolRef =
-        cudaq::opt::factory::createLLVMFunctionSymbol(
-            cudaq::opt::NVQIRPackSingleQubitInArray,
-            /*return type=*/qubitArrayType, {qubitIndexType}, parentModule);
-    // Pack the qubit into the array
-    Value result =
-        rewriter
-            .create<LLVM::CallOp>(loc, qubitArrayType, packingSymbolRef,
-                                  adaptor.getControls().front())
-            .getResult();
-    // The array result is what we want for the function args
-    funcArgs.push_back(result);
-
     // This is a single target op, add that type
     tmpArgTypes.push_back(qubitIndexType);
 
+    // __quantum__qis__NAME__ctl(double, Array*, Qubit*) Type
+    auto instOpQISFunctionType = LLVM::LLVMFunctionType::get(
+        LLVM::LLVMVoidType::get(context), tmpArgTypes);
+
+    // Get function pointer to ctrl operation
     FlatSymbolRefAttr instSymbolRef =
         cudaq::opt::factory::createLLVMFunctionSymbol(
             qirFunctionName, /*return type=*/LLVM::LLVMVoidType::get(context),
             std::move(tmpArgTypes), parentModule);
 
+    // Now we know we have the proper veq/ref type, we can
+    // handle multiple controls.
+    if (numControls > 1) {
+
+      // Get symbol for
+      // void invokeWithControlQubits(const std::size_t nControls, void
+      // (*QISFunction)(double, Array*, Qubit*), Qubit*, ...);
```

Comment:
This will still have an issue with mixed arg types (veq, ref, ref, veq, veq, etc...). We'll need an updated invokeWithControlQubits signature that provides info that the kth argument is an Array* or not.

Reply:
This is being handled in PR #876 (just commenting this for future reference -- I know you're the author of that PR haha)